### PR TITLE
AK: Introduce the new String, replacement for DeprecatedString

### PR DESCRIPTION
**Pull request by awesomekling**
DeprecatedString (formerly String) has been with us since the start, and it has served us well. However, it has a number of shortcomings that I'd like to address.

Some of these issues are hard if not impossible to solve incrementally inside of DeprecatedString, so instead of doing that, let's build a new String class and then incrementally move over to it instead.

Problems in DeprecatedString:

- It assumes string allocation never fails. This makes it impossible to use in allocation-sensitive contexts, and is the reason we had to ban DeprecatedString from the kernel entirely.

- The awkward null state. DeprecatedString can be null. It's different from the empty state, although null strings are considered empty. All code is immediately nicer when using Optional<DeprecatedString> but DeprecatedString came before Optional, which is how we ended up like this.

- The encoding of the underlying data is ambiguous. For the most part, we use it as if it's always UTF-8, but there have been cases where we pass around strings in other encodings (e.g ISO8859-1)

- operator[] and length() are used to iterate over DeprecatedString one byte at a time. This is done all over the codebase, and will *not* give the right results unless the string is all ASCII.

How we solve these issues in the new String:

- Functions that may allocate now return ErrorOr<String> so that ENOMEM errors can be passed to the caller.

- String has no null state. Use Optional<String> when needed.

- String is always UTF-8. You *can* jam bogus data in there if you really want to, but the API tries its best to ensure you don't.

- There is no operator[] or length(). You can get the underlying data with bytes(), but for iterating over code points, you should be using an UTF-8 iterator.

Furthermore, it has two nifty new features:

- String implements a small string optimization (SSO) for strings that can fit entirely within a pointer. This means up to 3 bytes on 32-bit platforms, and 7 bytes on 64-bit platforms. Such small strings will not be heap-allocated.

- String can create substrings without making a deep copy of the substring. Instead, the superstring gets +1 refcount from the substring, and it acts like a view into the superstring. To make substrings like this, use the substring_with_shared_superstring() API.

One caveat:

- String does not guarantee that the underlying data is null-terminated like DeprecatedString does today. While this was nifty in a handful of places where we were calling C functions, it did stand in the way of shared-superstring substrings.

### CODE REVIEW

File: AK/String.cpp [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+}
+
+String::String()
+{
+    m_short_string.byte_count_and_type = 0x01;
+}
+
+ErrorOr<String> String::from_utf8(StringView view)
+{
+    if (view.length() <= MAX_SHORT_STRING_LENGTH) {
```
Comment (linusg) [inline AK/String.cpp:126]: Let's not repeat the same mistakes we're tring to fix here :)

`MAX_SHORT_STRING_BYTE_COUNT`

File: AK/String.cpp [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+        ShortString short_string;
+        if (!view.is_empty())
+            memcpy(short_string.storage, view.characters_without_null_termination(), view.length());
+        short_string.byte_count_and_type = (view.length() << 1) | 0x01;
+        return String { short_string };
+    }
+    auto data = TRY(StringData::from_utf8(view.characters_without_null_termination(), view.length()));
+    return String { data };
```
Comment (linusg) [inline AK/String.cpp:134]:
```suggestion
    return String { move(data) };
```

File: AK/String.cpp [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        return m_data == other.m_data;
+    return bytes_as_string_view() == other.bytes_as_string_view();
+}
+
+bool String::operator==(StringView other) const
+{
+    return bytes_as_string_view() == other;
+}
+
+bool String::operator<(String const& other) const
```
Comment (linusg) [inline AK/String.cpp:173]: I'd be wary about introducing `operator<` and `operator>` as they don't work for general alphabetical sorting of unicode strings (think a/ã/ä/â/...), instead we should have (already might have?) something like `localeCompare` in JS.

File: AK/String.cpp [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+{
+    return bytes_as_string_view() < other.bytes_as_string_view();
+}
+
+bool String::operator>(String const& other) const
+{
+    return bytes_as_string_view() > other.bytes_as_string_view();
+}
+
+ErrorOr<String> String::substring(size_t start, size_t byte_count) const
```
Comment (linusg) [inline AK/String.cpp:183]: I feel like the function name should make it clearer that you're creating a substring of N bytes, not code points (Jakt solves this by mandatory parameter names, but that's N/A in C++)

File: AK/String.cpp [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+DeprecatedString String::to_deprecated_string() const
+{
+    return DeprecatedString(bytes_as_string_view());
+}
+
+ErrorOr<String> String::from_deprecated_string(DeprecatedString const& deprecated_string)
+{
+    Utf8View view(deprecated_string);
+    if (!view.validate())
+        return Error::from_string_literal("String::from_deprecated_string: Input was not valid UTF-8");
```
Comment (linusg) [inline AK/String.cpp:334]: Neither `StringData::from_utf8()` nor `String::from_utf8()` seems to guarantee the validity of the provided UTF-8 data, unless I'm missing something. Should we really do this (only) here then?

File: AK/String.h [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    {
+        VariadicFormatParams variadic_format_parameters { parameters... };
+        return vformatted(fmtstr.view(), variadic_format_parameters);
+    }
+
+    // NOTE: This is primarily interesting to unit tests.
+    [[nodiscard]] bool is_short_string() const;
+
+    // FIXME: Remove these once all code has been ported to String
+    DeprecatedString to_deprecated_string() const;
```
Comment (linusg) [inline AK/String.h:105]: `[[nodiscard]]`

File: Tests/AK/TestString.cpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    auto shorty = MUST(String::from_utf8("abc"sv));
+    EXPECT_EQ(shorty.is_short_string(), true);
+    EXPECT_EQ(shorty.bytes().size(), 3u);
+    EXPECT_EQ(shorty.bytes_as_string_view(), "abc"sv);
+#endif
+}
+
+TEST_CASE(long_strings)
+{
+    auto shorty = MUST(String::from_utf8("abcdefgh"sv));
```
Comment (linusg) [inline Tests/AK/TestString.cpp:44]: name doesn't seem to match the test case here :)

**Review by linusg — CHANGES_REQUESTED** [review]

File: AK/String.cpp (hunk 5 above)
Comment (linusg) [reply AK/String.cpp:334]: As in, if the new string is not happy to fully commit to UTF-8 validity, it should not assume that from the old string either.

File: AK/String.cpp (hunk 3 above)
Comment (linusg) [reply AK/String.cpp:173]: And if you think we should have it, an explicit name ("bytes lower than"?) might be a better choice.

File: AK/String.cpp (hunk 5 above)
Comment (awesomekling) [reply AK/String.cpp:334]: Let's commit.

**Review by linusg — APPROVED** [review]
LGTM!

**Pull request merged**